src/components/building/building.js

```diff
@@ -0,0 +1,115 @@
+import {buildingPrice} from '../../util/BuildingPrice';
+import {buildingTime, timeBuilder} from '../../util/BuildingTime';
+import Buildingcard from './buildingcard';
+import {bindActionCreators} from 'redux';
+import {useSelector, useDispatch} from 'react-redux';
+import {actionCreators} from '../../state';
+import {gameBuildingsTypes} from '../../util/gamedatas/gameBuildingsTypes';
+
+export const Building = () => {
+  const {removeResources, addBuildingToBuild} = bindActionCreators(actionCreators, useDispatch());
+
+  const currentUserResources = useSelector(state => state.currentUserResources);
+  const currentUserBuildings = useSelector(state => state.currentUserBuildings);
+  const currentUserBuildingInProgress = useSelector(state => state.currentUserBuildingInProgress);
+
+  function onHandleClickUpgrade(e) {
+    const buildingId = e.target.dataset.buildid;
+    const buildingBuildTime = e.target.dataset.buildtime;
+
+    if (currentUserBuildingInProgress.length === 0) {
+      const currentBuilding = currentUserBuildings.find(currentBuilding => currentBuilding.buildingId === buildingId);
+      const nextBuildingLevel = currentBuilding.level + 1;
+      const progressBuildingTime = buildingBuildTime * 1000;
+
+      const startBuildingTime = Date.now();
+      const endBuildingTime = startBuildingTime + progressBuildingTime;
+
+      addBuildingToBuild(0, buildingId, startBuildingTime, endBuildingTime, progressBuildingTime, nextBuildingLevel);
+
+      const buildingType = gameBuildingsTypes.find(buildingType => buildingType.id === buildingId);
+
+      buildingType.buildMaterials.forEach(material => {
+        if (material.resourceType !== 'energy') {
+          removeResources(
+            material.resourceType,
+            buildingPrice(nextBuildingLevel, buildingType.id, material.resourceType)
+          );
+        }
+      });
+    }
+  }
+
+  return gameBuildingsTypes.map(gameBuildingType => {
+    let currentUserBuilding = currentUserBuildings.find(
+      userBuilding => userBuilding.buildingId === gameBuildingType.id
+    );
+
+    let inProgressId = 0;
+    let buttonTexts = 'Build';
```

Comment:
Again, these are anti-patterns in React and this should be handled with a state. Even though you added redux to your app, for state updates within an app, I would recommend to use normal react state.

Reply:
I have removed the initialization of the let inProgressId = 0; and inserted it directly in the function as const